graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResponseFactory.java

```diff
@@ -0,0 +1,130 @@
+    private static final String LICENSE = "license";
+    private static final String PLUGIN = "plugin";
+    private static final String SEARCH_CLUSTER = "search_cluster";
+    private final ObjectMapper objectMapper;
+
+    @Inject
+    public TelemetryResponseFactory(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+    }
+
+    Map<String, Object> createTelemetryResponse(Map<String, Object> clusterInfo,
+                                                Map<String, Object> userInfo,
+                                                Map<String, Object> pluginInfo,
+                                                Map<String, Object> searchClusterInfo,
+                                                List<TelemetryLicenseStatus> licenseStatuses,
+                                                TelemetryUserSettings telemetryUserSettings) {
+        Map<String, Object> telemetryResponse = new LinkedHashMap<>();
+        telemetryResponse.put(CURRENT_USER, userInfo);
+        telemetryResponse.put(USER_TELEMETRY_SETTINGS, telemetryUserSettings);
+        telemetryResponse.put(CLUSTER, clusterInfo);
+        telemetryResponse.put(LICENSE, createLicenseInfo(licenseStatuses));
+        telemetryResponse.put(PLUGIN, pluginInfo);
+        telemetryResponse.put(SEARCH_CLUSTER, searchClusterInfo);
+        return telemetryResponse;
+    }
+
+    Map<String, Object> createTelemetryDisabledResponse(TelemetryUserSettings telemetryUserSettings) {
+        return Map.of(USER_TELEMETRY_SETTINGS, telemetryUserSettings);
+    }
+
+    Map<String, Object> createUserInfo(String userHash,
+                                       boolean isLocalAdmin,
+                                       int rolesCount,
+                                       int teamsCount) {
+        Map<String, Object> userInfo = new LinkedHashMap<>();
+        userInfo.put("user", userHash);
+        userInfo.putAll(flatten(Map.of(CURRENT_USER, Map.of(
+                "is_local_admin", isLocalAdmin,
+                "roles_count", rolesCount,
+                "teams_count", teamsCount
+
+        ))));
+        return userInfo;
+    }
+
+    Map<String, Object> createClusterInfo(String clusterId,
+                                          DateTime clusterCreationDate,
+                                          int nodesCount,
+                                          Map<String, SystemOverviewResponse> nodes,
+                                          long averageLastMonthTraffic,
+                                          long usersCount) {
+        Map<String, Object> clusterInfo = new LinkedHashMap<>();
+        clusterInfo.put("cluster_id", clusterId);
+        clusterInfo.put(f("%s.cluster_creation_date", CLUSTER), clusterCreationDate);
+        clusterInfo.put(f("%s.nodes_count", CLUSTER), nodesCount);
+        clusterInfo.put(f("%s.average_last_month_traffic", CLUSTER), averageLastMonthTraffic);
+        clusterInfo.put(f("%s.users_count", CLUSTER), usersCount);
```

Comment:
I think this does not need the `%s`, `CLUSTER` nesting, we will only ever have a single cluster here. If we ever have multi-cluster setups that users interact with, that will need a different representation anyay.
Remove the cluster id here makes the reported properties much simpler, too.

Reply:
As discussed, the cluster node data is no longer flattened.